atf-sh/atf-sh.1
.\"
.\" Automated Testing Framework (atf)
.\"
.\" Copyright (c) 2010 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND
.\" CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
.\" GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
.\" IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
.\" IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd September 27, 2014
.Dt ATF-SH 1
.Os
.Sh NAME
.Nm atf-sh
.Op Fl s Ar shell
.Nd interpreter for shell-based test programs
.Sh SYNOPSIS
.Nm
.Ar script
.Sh DESCRIPTION
.Nm
is an interpreter that runs the test program given in
.Ar script
after loading the
.Xr atf-sh 3
library.
.Pp
.Nm
is not a real interpreter though: it is just a wrapper around
the system-wide shell defined by
.Va ATF_SHELL .
.Nm
executes the interpreter, loads the
.Xr atf-sh 3
library and then runs the script.
You must consider
.Nm atf-sh
to be a POSIX shell by default and thus should not use any non-standard
extensions.
.Pp
The following options are available:
.Bl -tag -width XsXshellXXX
.It Fl s Ar shell
Specifies the shell to use instead of the value provided by
.Va ATF_SHELL .
.El
.Sh ENVIRONMENT
.Bl -tag -width ATFXLIBEXECDIRXX -compact
.It Va ATF_LIBEXECDIR
Overrides the builtin directory where
.Nm
is located.
Should not be overridden other than for testing purposes.
.It Va ATF_PKGDATADIR
Overrides the builtin directory where
.Pa libatf-sh.subr
is located.
Should not be overridden other than for testing purposes.
.It Va ATF_SHELL
Path to the system shell to be used in the generated scripts.
Scripts must not rely on this variable being set to select a specific
interpreter.
.El
.Sh EXAMPLES
Scripts using
.Xr atf-sh 3
should start with:
.Bd -literal -offset indent
#! /usr/bin/env atf-sh
.Ed
.Pp
Alternatively, if you want to explicitly choose a shell interpreter, you cannot
rely on
.Xr env 1
to find
.Nm .
Instead, you have to hardcode the path to
.Nm
in the script and then use the
.Fl s
option afterwards as a
.Em single parameter :
.Bd -literal -offset indent
#! /path/to/bin/atf-sh -s/bin/bash
.Ed
.Sh SEE ALSO
.Xr atf-sh 3


NEWS
Major changes between releases                  Automated Testing Framework
===========================================================================

Changes in version 0.21
***********************

STILL UNDER DEVELOPMENT; NOT RELEASED YET.

* Restored the atf(7) manual page to serve as a reference to all the other
  manual pages shipped by ATF.

* Added the -s flag to atf-sh to support specifying the shell interpreter
  to be used.

* Removed ATF_WORKDIR.  The only remaining consumers have been converted to
  use the standard TMPDIR environment variable.  As a benefit, and because
  Kyua forces the TMPDIR to live within the test case's work directory,
  any stale files left behind by ATF will be automatically cleaned up.

* Documented the environment variables recognized by each component in the
  relevant manual pages.  This information was lost with the atf-config(1)
  removal.

* Added a new "require.diskspace" metadata property to test cases so that
  they can specify the minimum amount of disk space required for the test
  to run.

* Renamed the atf-{c,c++,sh}-api(3) manual pages to atf-{c,c++,sh}(3) for
  discoverability purposes.  Symbolic links are provided for the time
  being to still make the old names visible.

* Issue #5: Recommend the (expected, actual) idiom for calls to the test
  macros in the manual pages.

* Issue #7: Stopped catching unhandled exceptions in atf-c++ tests.  This
  propagates the crash to the caller, which in turn allows it to obtain
  proper debugging information.  In particular, Kyua should now be able to
  extract a stacktrace pinpointing the problem.

* Issue #8: Fixed atf-c/macros_test:use test failures spotted by the clang
  that ships with FreeBSD 11.0-CURRENT.

* Issue #12: Improved documentation of atf-sh-api(3) and atf-check(1) by
  better explaining how they relate to each other.

* Issue #14: Stop setting 'set -e' in atf-sh.  This setting was initially
  added as a way to enable a "strict" mode in the library and to make test
  cases fail fast when they run unprotected commands.  However, doing so
  in the library is surprising as the responsibility of enabling 'set -e'
  should be on the user's code.  Also, 'set -e' introduces inconsistent
  behavior on subshells and users do not expect that.

* Issue #15: Fix atf_utils_{fork,wait} to support nested calls.

* Issue #16: Fixed test failures (by removing a long-standing hack) on
  systems that lack \e support in printf(1).

* Issue #19: Removed stale references to atf-config and atf-run.


Changes in version 0.20
***********************

Experimental version released on February 7th, 2014.

This is the first release without the code for the deprecated tools.  If
you require such code, please fetch a copy of the 0.19 release and extract
the 'tools' directory for your own consumption.

* Removed the deprecated tools.  This includes atf-config, atf-report,
  atf-run and atf-version.


Changes in version 0.19
***********************

Experimental version released on February 7th, 2014.

This is the last release to bundle the code for the deprecated tools.
The next release will drop their code and will stop worrying about
backwards compatibility between the ATF libraries and what the old tools
may or may not support.

If you still require the old tools for some reason, grab a copy of the
'tools' directory now.  The code in this directory is standalone and
does not depend on any internal details of atf-c++ any longer.

* Various fixes and improvements to support running as part of the FreeBSD
  test suite.

* Project hosting moved from Google Code (as a subproject of Kyua) to
  GitHub (as a first-class project).  The main reason for the change is
  the suppression of binary downloads in Google Code on Jan 15th, 2014.
  See https://github.com/jmmv/atf/

* Removed builtin help from atf-sh(1) and atf-check(1) for simplicity
  reasons.  In other words, their -h option is gone.

* Moved the code of the deprecated tools into a 'tools' directory and
  completely decoupled their code from the internals of atf-c++.  The
  reason for this is to painlessly allow a third-party to maintain a
  copy of these tools after we delete them because upcoming changes to
  atf-c++ would break the stale tools.


Changes in version 0.18
***********************

Experimental version released on November 16th, 2013.

* Issue 45: Added require.memory support in atf-run for FreeBSD.

* Fixed an issue with the handling of cin with libc++.

* Issue 64: Fixed various mandoc formatting warnings.

* NetBSD PR bin/48284: Made atf-check flush its progress message to
  stdout so that an interrupted test case always shows the last message
  being executed.

* NetBSD PR bin/48285: Fixed atf_check examples in atf-sh-api(3).


Changes in version 0.17
***********************

Experimental version released on February 14th, 2013.

* Added the atf_utils_cat_file, atf_utils_compare_file,
  atf_utils_copy_file, atf_utils_create_file, atf_utils_file_exists,
  atf_utils_fork, atf_utils_grep_file, atf_utils_grep_string,
  atf_utils_readline, atf_utils_redirect and atf_utils_wait utility
  functions to atf-c-api.  Documented the already-public
  atf_utils_free_charpp function.

* Added the cat_file, compare_file, copy_file, create_file, file_exists,
  fork, grep_collection, grep_file, grep_string, redirect and wait
  functions to the atf::utils namespace of atf-c++-api.  These are
  wrappers around the same functions added to the atf-c-api library.

* Added the ATF_CHECK_MATCH, ATF_CHECK_MATCH_MSG, ATF_REQUIRE_MATCH and
  ATF_REQUIRE_MATCH_MSG macros to atf-c to simplify the validation of a
  string against a regular expression.

* Miscellaneous fixes for manpage typos and compilation problems with
  clang.

* Added caching of the results of those configure tests that rely on
  executing a test program.  This should help crossbuild systems by
  providing a mechanism to pre-specify what the results should be.


atf-sh/atf-sh.cpp
//
// Automated Testing Framework (atf)
//
// Copyright (c) 2010 The NetBSD Foundation, Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND
// CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
// IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
// GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
// IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//

extern "C" {
#include <unistd.h>
}

#include <cerrno>
#include <cstdlib>
#include <cstring>
#include <iostream>

#include "atf-c++/detail/application.hpp"
#include "atf-c++/detail/env.hpp"
#include "atf-c++/detail/fs.hpp"
#include "atf-c++/detail/sanity.hpp"

// ------------------------------------------------------------------------
// Auxiliary functions.
// ------------------------------------------------------------------------

namespace {

static
std::string
fix_plain_name(const char *filename)
{
    const atf::fs::path filepath(filename);
    if (filepath.branch_path().str() == ".")
        return std::string("./") + filename;
    else
        return std::string(filename);
}

static
std::string*
construct_script(const char* filename)
{
    const std::string libexecdir = atf::env::get(
        "ATF_LIBEXECDIR", ATF_LIBEXECDIR);
    const std::string pkgdatadir = atf::env::get(
        "ATF_PKGDATADIR", ATF_PKGDATADIR);
    const std::string shell = atf::env::get("ATF_SHELL", ATF_SHELL);

    std::string* command = new std::string();
    command->reserve(512);
    (*command) += ("Atf_Check='" + libexecdir + "/atf-check' ; " +
                   "Atf_Shell='" + shell + "' ; " +
                   ". " + pkgdatadir + "/libatf-sh.subr ; " +
                   ". " + fix_plain_name(filename) + " ; " +
                   "main \"${@}\"");
    return command;
}

static
const char**
construct_argv(const std::string& shell, const int interpreter_argc,
               const char* const* interpreter_argv)
{
    PRE(interpreter_argc >= 1);
    PRE(interpreter_argv[0] != NULL);

    const std::string* script = construct_script(interpreter_argv[0]);

    const int count = 4 + (interpreter_argc - 1) + 1;
    const char** argv = new const char*[count];
    argv[0] = shell.c_str();
    argv[1] = "-c";
    argv[2] = script->c_str();
    argv[3] = interpreter_argv[0];

    for (int i = 1; i < interpreter_argc; i++)
        argv[4 + i - 1] = interpreter_argv[i];

    argv[count - 1] = NULL;

    return argv;
}

} // anonymous namespace

// ------------------------------------------------------------------------
// The "atf_sh" class.
// ------------------------------------------------------------------------

class atf_sh : public atf::application::app {
    static const char* m_description;

    atf::fs::path m_shell;

    options_set specific_options(void) const;
    void process_option(int, const char*);

public:
    atf_sh(void);

    int main(void);
};

const char* atf_sh::m_description =
    "atf-sh is a shell interpreter that extends the functionality of the "
    "system sh(1) with the atf-sh library.";

atf_sh::atf_sh(void) :
    app(m_description, "atf-sh(1)"),
    m_shell(atf::fs::path(atf::env::get("ATF_SHELL", ATF_SHELL)))
{
}

atf_sh::options_set
atf_sh::specific_options(void)
    const
{
    using atf::application::option;
    options_set opts;

    INV(m_shell == atf::fs::path(atf::env::get("ATF_SHELL", ATF_SHELL)));
    opts.insert(option('s', "shell", "Path to the shell interpreter to use; "
                       "default: " + m_shell.str()));

    return opts;
}

void
atf_sh::process_option(int ch, const char* arg)
{
    switch (ch) {
    case 's':
        m_shell = atf::fs::path(arg);
        break;

    default:
        UNREACHABLE;
    }
}

int
atf_sh::main(void)
{
    if (m_argc < 1)
        throw atf::application::usage_error("No test program provided");

    const atf::fs::path script(m_argv[0]);
    if (!atf::fs::exists(script))
        throw std::runtime_error("The test program '" + script.str() + "' "
                                 "does not exist");

    const char** argv = construct_argv(m_shell.str(), m_argc, m_argv);
    // Don't bother keeping track of the memory allocated by construct_argv:
    // we are going to exec or die immediately.

    const int ret = execv(m_shell.c_str(), const_cast< char** >(argv));
    INV(ret == -1);
    std::cerr << "Failed to execute " << m_shell.str() << ": "
              << std::strerror(errno) << "\n";
    return EXIT_FAILURE;
}

int
main(int argc, char* const* argv)
{
    return atf_sh().run(argc, argv);
}
